src/charm.py

```diff
@@ -1,29 +1,130 @@
 #!/usr/bin/env python3
 # Copyright 2021 Canonical Ltd.
 # See LICENSE file for licensing details.
+import json
+import logging
 import textwrap
+import urllib.parse
+from typing import Any, Dict, List, Tuple, Optional
 
+import kubernetes
+import ops
+import requests
+import yaml
 from charms.alertmanager_k8s.v0.alertmanager import AlertmanagerProvider
 from charms.karma_k8s.v0.karma import KarmaConsumer
-import utils
-
-import ops
-from ops.charm import CharmBase, ActionEvent
-from ops.main import main
+from ops.charm import ActionEvent, CharmBase
 from ops.framework import StoredState
-from ops.model import ActiveStatus, MaintenanceStatus, BlockedStatus
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
 from ops.pebble import ChangeError
 
-import urllib.parse
-import requests
-import yaml
-import json
-import logging
-from typing import Dict, List, Optional, Any
+import utils
 
 logger = logging.getLogger(__name__)
 
 
+class K8sServicePatch:
+    namespace_file = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+
+    class PatchFailed(RuntimeError):
```

Comment:
I don't see why you would like an exception type to be a nested class.

Reply:
`class K8sServicePatch` is essentially just a namespace for related stuff, that's why the nesting.
The "PatchFailed" exception makes no sense anywhere else outside.
Now that I'll be moving it to its own file/module (i.e. namespace) I will take it out.